openLRSng/TX.h

```diff
@@ -26,83 +19,455 @@ uint8_t RX_ain0 = 0;
+    }
+    tx_buf[0] |= (0x37 + serial_resend[0]);
+    serialSendState = SERIAL_ACK; // sent but not acked
 
-static inline void processPulse(uint16_t pulse)
+  } else {
+    uint16_t PPMout[16];
+    if (failsafeState == 2) {
+      tx_buf[0] |= 0x01; // save failsafe
+      Red_LED_ON
+    } else {
+      tx_buf[0] |= 0x00; // servo positions
+      Red_LED_OFF
+      if (serialSendState == SERIAL_WAIT0) {
+        serialSendState = SERIAL_WAIT1;
+      }
+      if (serialSendState == SERIAL_ACK) {
+        serialSendState = SERIAL_RESEND;  // resend
+      }
+    }
+    for (uint8_t i = 0; i < PPM_CHANNELS; i++) {
+      PPMout[i] = getChannel(i);
+    }
+    packChannels((bind_data.flags & 0x07), PPMout, (tx_buf + 1));
+  }
+
+  //Green LED will be on during transmission
+  Green_LED_ON;
+  power = bind_data.rf_power;
+
+#ifdef TX_MODE1
+  if (tx_config.flags & SW_POWER) {
+    if (!digitalRead(TX_MODE1)) {
+      power = 7;
+    }
+  }
+#endif
+
+  if (serialMode == SERIAL_MODE_MULTI) {
+    power = checkMultiPower(power);
+  }
+
+  rfmSetPower(power);
+
+  //Hop to the next frequency
+  RF_channel++;
+  if ((RF_channel == MAXHOPS) || (bind_data.hopchannel[RF_channel] == 0)) {
+    RF_channel = 0;
+  }
+  setHopChannel(RF_channel);
+
+  // Send the data over RF
+  tx_packet_async(tx_buf, getPacketSize(&bind_data));
+  Green_LED_OFF;
+}
+
+void handlePacketTelem(void)
+{
+  if (RF_Mode != RECEIVED) {
+    return;
+  }
+
+  // got telemetry packet
+  lastTelemetry = lastPktTimeRX;
+  linkQualityTX |= 1;
+
+  rfmGetPacket(rx_buf, DOWNLINK_PACKETSIZE);
+
+  if ((tx_buf[0] ^ rx_buf[0]) & DOWNLINK_SEQBIT) {
+    // swap sequence to ACK
+    // received data from RX
+    tx_buf[0] ^= DOWNLINK_SEQBIT;
+
+    if ((rx_buf[0] & 0x38) == 0x38) {
+      // transparent serial data...
+      for (uint8_t i = 0; i <= (rx_buf[0] & 0x07);) {
+        if (bind_data.flags & TELEMETRY_FRSKY) {
+          frskyUserData(rx_buf[++i]);
+          if (serialMode == SERIAL_MODE_MULTI) {
+            sendMultiTelemetry(true, true, false);
+          }
+        } else {
+          rcSerial->write(rx_buf[++i]);
+        }
+      }
+    } else if ((rx_buf[0] & 0x3F) == 0) {
+      RSSI_rx = rx_buf[1];
+      RX_ain0 = rx_buf[2];
+      RX_ain1 = rx_buf[3];
+      linkQualityRX = rx_buf[6];
+    }
+  }
+
+  if (serialSendState == SERIAL_WAIT1) {
+    serialSendState = SERIAL_SEND;
+  } else if (serialSendState == SERIAL_ACK) {
+    serialSendState = SERIAL_WAIT0;
+  }
+}
+
+void checkTelemLinkState(void)
+{
+  if (!lastTelemetry) {
+    return;
+  }
+
+  if ((loopTime - lastTelemetry) > hopInterval) {
+    // telemetry lost
+    if (!(tx_config.flags & MUTE_TX)) {
+      buzzerOn(BZ_FREQ);
+    }
+    lastTelemetry = 0;
+  } else {
+    // telemetry link re-established
+    buzzerOff();
+  }
+}
+
+void checkFrSkyTelem(void)
+{
+  if (lastTelemetry && (bind_data.flags & TELEMETRY_FRSKY)) {
+    uint8_t linkQualityTX = countSetBits(linkQualityTX & 0xFFfe);
```

Comment:
This seems to be wrong. Or at the very least it needs `linkQualityTX` to be initialised before being used.

Reply:
it should have been initialized to zero.
If not, I'll fix it.

Reply:
actually, that looks like a bug. will fix...